File: services/order-service/src/rabbit.ts
import amqp, { Channel, ChannelModel, ConsumeMessage } from "amqplib";
import { v4 as uuidv4 } from "uuid";

import { logger } from "./logger";

export const EXCHANGE = "food_delivery";
export const RETRY_EXCHANGE = "food_delivery.retry";
export const DLX = "food_delivery.dlx";
export const MAX_RETRIES = 3;
export const RETRY_TTL_MS = 30_000;
export const PREFETCH = 10;

export type EventType =
  | "order.placed"
  | "order.accepted"
  | "order.rejected"
  | "order.ready"
  | "order.picked_up"
  | "order.delivered"
  | "order.cancelled"
  | "payment.completed"
  | "payment.failed"
  | "delivery.assigned";

export interface Envelope<T = unknown> {
  event_id: string;
  event_type: EventType;
  event_version: string;
  occurred_at: string;
  producer: string;
  data: T;
}

export class Rabbit {
  private conn!: ChannelModel;
  private ch!: Channel;

  constructor(
    private url: string,
    private producer: string,
    private serviceQueueBase: string, // e.g. "order"
  ) {}

  async connect(): Promise<void> {
    this.conn = await amqp.connect(this.url);
    this.ch = await this.conn.createChannel();
    await this.ch.prefetch(PREFETCH);

    await this.ch.assertExchange(EXCHANGE, "topic", { durable: true });
    await this.ch.assertExchange(RETRY_EXCHANGE, "direct", { durable: true });
    await this.ch.assertExchange(DLX, "fanout", { durable: true });

    const main = `${this.serviceQueueBase}.events`;
    const retry = `${this.serviceQueueBase}.events.retry`;
    const dlq = `${this.serviceQueueBase}.events.dlq`;

    await this.ch.assertQueue(main, { durable: true });
    await this.ch.assertQueue(retry, {
      durable: true,
      arguments: {
        "x-message-ttl": RETRY_TTL_MS,
        "x-dead-letter-exchange": RETRY_EXCHANGE,
        "x-dead-letter-routing-key": main,
      },
    });
    await this.ch.assertQueue(dlq, { durable: true });

    // Retry exchange routes back to main queue so retried messages are reprocessed.
    await this.ch.bindQueue(main, RETRY_EXCHANGE, main);
    await this.ch.bindQueue(dlq, DLX, "");

    this.conn.on("error", (err) => logger.error({ err }, "rabbit connection error"));
  }

  async publish<T>(event: EventType, data: T): Promise<void> {
    const envelope: Envelope<T> = {
      event_id: uuidv4(),
      event_type: event,
      event_version: "1.0",
      occurred_at: new Date().toISOString(),
      producer: this.producer,
      data,
    };
    this.ch.publish(EXCHANGE, event, Buffer.from(JSON.stringify(envelope)), {
      persistent: true,
      contentType: "application/json",
      messageId: envelope.event_id,
    });
  }

  /**
   * Subscribe to one or more routing keys on the service's main queue.
   * Handler receives the parsed envelope. Throwing routes to retry/DLQ.
   */
  async subscribe(
    routingKeys: EventType[],
    handler: (env: Envelope) => Promise<void>,
  ): Promise<void> {
    const main = `${this.serviceQueueBase}.events`;
    for (const k of routingKeys) {
      await this.ch.bindQueue(main, EXCHANGE, k);
    }
    await this.ch.consume(main, async (msg) => {
      if (!msg) return;
      try {
        const env = JSON.parse(msg.content.toString()) as Envelope;
        await handler(env);
        this.ch.ack(msg);
      } catch (err) {
        const retries = countRetries(msg);
        if (retries >= MAX_RETRIES) {
          logger.error({ err, retries }, "max retries exceeded → DLQ");
          this.ch.publish(DLX, "", msg.content, {
            persistent: true,
            contentType: "application/json",
            messageId: msg.properties.messageId,
          });
          this.ch.ack(msg);
        } else {
          logger.warn({ err, retries }, "consumer error → retry");
          this.ch.publish(RETRY_EXCHANGE, `${this.serviceQueueBase}.events.retry`, msg.content, {
            persistent: true,
            contentType: "application/json",
            messageId: msg.properties.messageId,
            headers: { "x-retry-count": retries + 1 },
          });
          // Route through retry queue for the TTL delay.
          await this.routeViaRetryQueue(msg);
          this.ch.ack(msg);
        }
      }
    });
  }

  private async routeViaRetryQueue(msg: ConsumeMessage): Promise<void> {
    const retry = `${this.serviceQueueBase}.events.retry`;
    this.ch.sendToQueue(retry, msg.content, {
      persistent: true,
      contentType: "application/json",
      messageId: msg.properties.messageId,
      headers: { ...(msg.properties.headers ?? {}) },
    });
  }

  async close(): Promise<void> {
    try {
      await this.ch?.close();
      await this.conn?.close();
    } catch {
      // ignore
    }
  }
}

function countRetries(msg: ConsumeMessage): number {
  const xDeath = msg.properties.headers?.["x-death"] as Array<{ count: number }> | undefined;
  if (Array.isArray(xDeath) && xDeath.length > 0) {
    return Number(xDeath[0]?.count ?? 0);
  }
  const explicit = msg.properties.headers?.["x-retry-count"];
  return typeof explicit === "number" ? explicit : 0;
}

File: services/payment-service/internal/events/rabbit.go
package events

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"time"

	"github.com/food-delivery/payment-service/internal/observability"
	"github.com/google/uuid"
	amqp "github.com/rabbitmq/amqp091-go"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

var tracer = otel.Tracer("payment-service/events")

const (
	Exchange      = "food_delivery"
	RetryExchange = "food_delivery.retry"
	DLX           = "food_delivery.dlx"
	MaxRetries    = 3
	RetryTTLMs    = 30_000
	Prefetch      = 10
)

type Envelope struct {
	EventID      string          `json:"event_id"`
	EventType    string          `json:"event_type"`
	EventVersion string          `json:"event_version"`
	OccurredAt   string          `json:"occurred_at"`
	Producer     string          `json:"producer"`
	Data         json.RawMessage `json:"data"`
}

type Rabbit struct {
	conn       *amqp.Connection
	ch         *amqp.Channel
	producer   string
	queueBase  string // e.g. "payment"
	mainQueue  string
	retryQueue string
	dlqQueue   string
}

func New(url, producer, queueBase string) (*Rabbit, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, fmt.Errorf("dial: %w", err)
	}
	ch, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, fmt.Errorf("channel: %w", err)
	}
	if err := ch.Qos(Prefetch, 0, false); err != nil {
		conn.Close()
		return nil, fmt.Errorf("qos: %w", err)
	}

	r := &Rabbit{
		conn:       conn,
		ch:         ch,
		producer:   producer,
		queueBase:  queueBase,
		mainQueue:  queueBase + ".events",
		retryQueue: queueBase + ".events.retry",
		dlqQueue:   queueBase + ".events.dlq",
	}
	if err := r.declareTopology(); err != nil {
		r.Close()
		return nil, err
	}
	return r, nil
}

func (r *Rabbit) declareTopology() error {
	if err := r.ch.ExchangeDeclare(Exchange, "topic", true, false, false, false, nil); err != nil {
		return err
	}
	if err := r.ch.ExchangeDeclare(RetryExchange, "direct", true, false, false, false, nil); err != nil {
		return err
	}
	if err := r.ch.ExchangeDeclare(DLX, "fanout", true, false, false, false, nil); err != nil {
		return err
	}
	if _, err := r.ch.QueueDeclare(r.mainQueue, true, false, false, false, nil); err != nil {
		return err
	}
	if _, err := r.ch.QueueDeclare(r.retryQueue, true, false, false, false, amqp.Table{
		"x-message-ttl":             int32(RetryTTLMs),
		"x-dead-letter-exchange":    RetryExchange,
		"x-dead-letter-routing-key": r.mainQueue,
	}); err != nil {
		return err
	}
	if _, err := r.ch.QueueDeclare(r.dlqQueue, true, false, false, false, nil); err != nil {
		return err
	}
	if err := r.ch.QueueBind(r.mainQueue, r.mainQueue, RetryExchange, false, nil); err != nil {
		return err
	}
	if err := r.ch.QueueBind(r.dlqQueue, "", DLX, false, nil); err != nil {
		return err
	}
	return nil
}

func (r *Rabbit) Publish(ctx context.Context, eventType string, data any) error {
	dataJSON, err := json.Marshal(data)
	if err != nil {
		return err
	}
	env := Envelope{
		EventID:      uuid.NewString(),
		EventType:    eventType,
		EventVersion: "1.0",
		OccurredAt:   time.Now().UTC().Format(time.RFC3339Nano),
		Producer:     r.producer,
		Data:         dataJSON,
	}
	body, err := json.Marshal(env)
	if err != nil {
		return err
	}
	ctx, span := tracer.Start(ctx, "amqp.publish "+eventType,
		trace.WithSpanKind(trace.SpanKindProducer),
		trace.WithAttributes(
			semconv.MessagingSystemRabbitmq,
			attribute.String("messaging.destination.name", Exchange),
			attribute.String("messaging.rabbitmq.routing_key", eventType),
			attribute.String("messaging.message.id", env.EventID),
		),
	)
	defer span.End()

	pub := amqp.Publishing{
		ContentType:  "application/json",
		DeliveryMode: amqp.Persistent,
		MessageId:    env.EventID,
		Timestamp:    time.Now().UTC(),
		Body:         body,
	}
	observability.InjectAMQP(ctx, &pub)
	return r.ch.PublishWithContext(ctx, Exchange, eventType, false, false, pub)
}

type Handler func(ctx context.Context, env *Envelope) error

func (r *Rabbit) Subscribe(routingKeys []string, handler Handler) error {
	for _, k := range routingKeys {
		if err := r.ch.QueueBind(r.mainQueue, k, Exchange, false, nil); err != nil {
			return err
		}
	}
	deliveries, err := r.ch.Consume(r.mainQueue, "", false, false, false, false, nil)
	if err != nil {
		return err
	}
	go r.consumeLoop(deliveries, handler)
	return nil
}

func (r *Rabbit) consumeLoop(deliveries <-chan amqp.Delivery, handler Handler) {
	for d := range deliveries {
		var env Envelope
		if err := json.Unmarshal(d.Body, &env); err != nil {
			slog.Error("malformed envelope, dropping", "err", err)
			_ = d.Nack(false, false)
			continue
		}
		parentCtx := observability.ExtractAMQP(context.Background(), d)
		ctx, cancel := context.WithTimeout(parentCtx, 30*time.Second)
		ctx, span := tracer.Start(ctx, "amqp.consume "+env.EventType,
			trace.WithSpanKind(trace.SpanKindConsumer),
			trace.WithAttributes(
				semconv.MessagingSystemRabbitmq,
				attribute.String("messaging.destination.name", r.mainQueue),
				attribute.String("messaging.message.id", env.EventID),
			),
		)
		err := handler(ctx, &env)
		span.End()
		cancel()
		if err == nil {
			_ = d.Ack(false)
			continue
		}
		retries := countRetries(d)
		if retries >= MaxRetries {
			slog.Error("max retries exceeded → DLQ", "event_id", env.EventID, "retries", retries, "err", err)
			_ = r.ch.PublishWithContext(context.Background(), DLX, "", false, false, amqp.Publishing{
				ContentType:  "application/json",
				DeliveryMode: amqp.Persistent,
				MessageId:    env.EventID,
				Body:         d.Body,
			})
			_ = d.Ack(false)
			continue
		}
		slog.Warn("consumer error → retry", "event_id", env.EventID, "retries", retries, "err", err)
		// retry queue → TTL expires → dead-letters back to main queue.
		_ = r.ch.PublishWithContext(context.Background(), "", r.retryQueue, false, false, amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			MessageId:    env.EventID,
			Body:         d.Body,
			Headers:      amqp.Table{"x-retry-count": int32(retries + 1)},
		})
		_ = d.Ack(false)
	}
}

func countRetries(d amqp.Delivery) int {
	if v, ok := d.Headers["x-retry-count"]; ok {
		switch n := v.(type) {
		case int32:
			return int(n)
		case int64:
			return int(n)
		case int:
			return n
		}
	}
	if xd, ok := d.Headers["x-death"]; ok {
		if list, ok := xd.([]interface{}); ok && len(list) > 0 {
			if entry, ok := list[0].(amqp.Table); ok {
				if c, ok := entry["count"]; ok {
					switch n := c.(type) {
					case int32:
						return int(n)
					case int64:
						return int(n)
					}
				}
			}
		}
	}
	return 0
}

func (r *Rabbit) Close() {
	if r.ch != nil {
		_ = r.ch.Close()
	}
	if r.conn != nil {
		_ = r.conn.Close()
	}
}
